### DISCUSSION TITLE
Monorepos with Firebase App Hosting

### DISCUSSION THREAD

**Discussion by benyap** [firebase/firebase-tools#7206] [General]
The new [Firebase App Hosting](https://firebase.google.com/docs/app-hosting) feature is really exciting!

Has anyone managed to get it working with monorepos? E.g. deploy a Next.js app to App Hosting, where the app uses code from other packages in the monorepo. In my testing, the pipeline is able to build my app, but fails when the Nextjs adapter does some post-build work with this error: `Error: Output directory is not of expected structure`

I'm assuming that it's because monorepos are not officially supported yet.

Would love to hear if anyone else has tried this, and if anyone knows if it's on the roadmap!

**Comment by bkendall** [comment]
In addition to your note here, you should put this request on [Firebase App Hosting's UserVoice page](https://firebase.uservoice.com/forums/948424-general?category_id=501599). We've been collecting a bunch of good requests there, and the voting there can help us figure out what to work on next!

**Reply by benyap** [reply]
Great idea! Will do.

**Comment by tolypash** [comment]
Hey! I'm getting the same issue!
Any progress @benyap ?

**Reply by benyap** [reply]
Glad to know I'm not the only one running into this! Waiting to see if the Firebase folks can work some magic on their side to make it happen. For now, I'm going to stick with a custom Docker image on Cloud Run.

**Comment by alexdabast** [comment]
Funny enough I was able to make app hosting working with my mono repo angular SSR app few months ago
Then firebase probably changed something because few days after it was not working anymore
![image](https://github.com/user-attachments/assets/82ac080e-5953-416d-a3f6-a31b0beb6f7f)

**Comment by tonyjhuang** [comment]
Hey all, Firebaser here. I'm excited to announce that we just landed our first monorepo support on App Hosting! You can now deploy Angular and Next.js projects managed by Nx by specifying the app directory as the `root_directory` for your backend.

We launched a guide to help you get started here: https://firebase.google.com/docs/app-hosting/monorepos

Thanks for your interest in App Hosting and as always feel free to vote on what features you'd like to see on [User Voice](https://firebase.uservoice.com/forums/948424-general?category_id=501599). That helps us prioritize what to focus on :)

**Reply by benyap** [reply]
Hi @tonyjhuang, appreciate the update!

This doesn't seem to be working for me. The guide looks promising but is a little lacking in detail for someone who isn't familiar with Nx, as I use `pnpm` workspaces with Turbo.

I followed this [setup guide](https://nx.dev/recipes/adopting-nx/adding-to-monorepo) to add Nx to my `pnpm` workspace. However, I'm unable to deploy to App Hosting even though I supplied a `root_directory` when I set up the backend - see the error below.

![image](https://github.com/user-attachments/assets/614d5576-8020-42c9-b8dc-8c2acb51f17f)

The recommendation to specify a "default project" does not really help here, because I'd like to deploy multiple backends to App Hosting from my monorepo (i.e. if I specified project A in the "default project" of nx.json, how can I get project B to deploy?)

Perhaps the documentation could have more detail on what exactly App Hosting is expecting to "recognise" a monorepo? As far as I'm concerned, my project "uses" Nx (which really, according to the guide, looks to be just `pnpm` workspaces). But in reality, I have no idea what App Hosting is looking for to deploy an app from my monorepo.

**Questions**

1. Where would be the best place to get support for the issue I'm having above? A GitHub issue?
2. Are there plans to support monorepo set ups _not_ managed by Nx? Wouldn't it make more sense to support more "general" monorepo tooling like npm/yarn/pnpm workspaces?

**Reply by benyap** [reply]
I did spot https://github.com/firebase/firebase-tools/issues/7478 which seems to be what I'm also looking for too - will follow that for updates.

**Comment by Muneersahel** [comment]
I'm currently getting the same issue on angular ssr project running on angular v18.2.6, nx v19.8.0. Any solution?

**Reply by tonyjhuang** [reply]
Can you reach out to us at [Firebase Support](https://firebase.google.com/support)? That way we can take a look at your build logs and application structure.
